[storage] Add OTLP Scope metadata validation test (part of #7050)

SoumyaRaikwar opened the pull request:
## Summary
Part of #7050 - Adds OTLP Scope validation test for v2 Storage API.

## Changes
- `integration.go`: Added `testOTLPScopePreservation()` test + helper methods
- `otlp_scope_attributes.json`: OTLP fixture with InstrumentationScope metadata
- `otlp_span_links.json`: OTLP fixture with span links

## Why
Validates OTLP-specific features (InstrumentationScope name/version) not testable with v1 model. Tests run automatically on all storage backends.

Part of #7050

yurishkuro commented on internal/storage/integration/fixtures/traces/otlp_resource_attributes.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "resourceSpans": [
```
where are these used?

yurishkuro reviewed (commented):
It does not fix 7050. I just see a patchwork of different threads.

SoumyaRaikwar commented on internal/storage/v2/v1adapter/translator.go:
```diff
@@ -90,6 +90,11 @@ func V1TraceToOtelTrace(jTrace *model.Trace) ptrace.Traces {
 }
 
+// V1TraceFromOtelTrace converts a single OTLP trace to v1 model.Trace
+func V1TraceFromOtelTrace(otelTrace ptrace.Traces) *model.Trace {
```
You're absolutely right - this function just wraps the internal modelTraceFromOtelTrace without adding value.

SoumyaRaikwar commented:
@yurishkuro - Thanks for the feedback! Cleaned up unused code in latest commit.

**Question on scope:** This PR currently upgrades `writeTrace()` to v2 API and adds OTLP scope validation. 

To fully complete #7050, I'd need to also convert all 30 fixture files and update 7 methods from `*model.Trace` → `ptrace.Traces` .

Should I:
1.  Expand this PR to include full migration?
2.  Keep it focused and do full migration in follow-up PR(s)?

Happy to do either - just want to confirm approach before implementing!

yurishkuro commented:
The full scope is defined in the issue. I'm ok to merge incremental changes, but they must be labeled "part of" not "fixes" for the issue.

SoumyaRaikwar commented:
@yurishkuro - Thanks for clarifying! My apologies for it.
Updated PR to reflect this is "part of #7050" as an incremental change. Will submit follow-up PRs for fixture conversion and remaining method migration.

Manik2708 commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +633,119 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+	s.writeTrace(t, traces)
+	traceID := extractTraceID(t, traces)
+
+	var readTraces []*model.Trace
```
why the new test has old `model.Trace`? we want to upgrade the tests completely to use otlp traces. if new fixtures are added, why we need to convert them back to old models?

SoumyaRaikwar replied:
it is now resolved: the scope test is fully OTLP-native.

SoumyaRaikwar commented:
@yurishkuro @Manik2708 

Thanks for the feedback! Updated:

-  Removed unsupported custom attributes from span link fixtures
-  Added comprehensive assertions verifying each link's TraceID and SpanID
-  Scope test now validates full metadata (name + version)

Tests confirm OTLP structure preservation through v2 API for memory backend.
could you please review again

yurishkuro commented on internal/storage/integration/integration.go:
```diff
@@ -368,7 +369,9 @@ func (s *StorageIntegration) testFindTraces(t *testing.T) {
 			if !ok {
 				trace = s.getTraceFixture(t, traceFixture)
-				s.writeTrace(t, trace)
+				otelTraces := v1adapter.V1TraceToOtelTrace(trace)
```
why not do V1TraceToOtelTrace inside getTraceFixture?

yurishkuro commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +639,199 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+		found      bool
+	)
+
+	for i := 0; i < resourceSpans.Len() && !found; i++ {
```
use iterator functions from OTLP

SoumyaRaikwar replied:
Done! Now using `.All()` iterators throughout `normalizeOTLPTimestamps()`:
- `for _, rs := range resourceSpans.All()`
- `for _, ss := range rs.ScopeSpans().All()` 
- `for _, span := range ss.Spans().All()`

yurishkuro commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +639,199 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+
+	found := s.waitForCondition(t, func(t *testing.T) bool {
+		ctx := context.Background()
+		iter := s.TraceReader.GetTraces(ctx, tracestore.GetTraceParams{TraceID: traceID})
```
```suggestion
		iter := s.TraceReader.GetTraces(t.Context(), tracestore.GetTraceParams{TraceID: traceID})
```

SoumyaRaikwar replied:
done

SoumyaRaikwar commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +639,199 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+		return false
+	})
+
+	require.True(t, found, "Failed to retrieve OTLP trace with span links")
```
Refactored to extract `testOTLPHelper()` which now handles all common 
boilerplate.

SoumyaRaikwar commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +639,199 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+	return traces
+}
+
+func normalizeOTLPTimestamps(traces ptrace.Traces) {
```
done

SoumyaRaikwar commented on internal/storage/integration/integration.go:
```diff
@@ -627,6 +639,199 @@ func (s *StorageIntegration) insertThroughput(t *testing.T) {
+	targetStart := time.Now().Add(-time.Minute).UTC()
+	delta := targetStart.Sub(firstStart)
+
+	for i := 0; i < resourceSpans.Len(); i++ {
```
Combined into a single loop - now detects the first timestamp inline while iterating and adjusts all timestamps in one pass.

yurishkuro reviewed (commented):
I don't think we should be adding new tests. First, you are mixing two things in this PR, OTLP-enablement (incomplete) and specific OTLP segments validation.

I prefer we consolidate everything in `testGetTrace`. We need to migrate the framework to work in OTLP space (e.g. CompareTraces function) and refresh the existing fixtures to be in OTLP.

Then, because not all storage backends support fully structured capture of OTLP data (like scope & link attributes, although I am rather skeptical if they are still not captured as span attributes by the OTLP->Jaeger translator), we can isolate those specific features into separate fixtures (making testGetTrace a nested set of tests) and exclude some of them from some storage tests by name, similar to how some of the query tests are excluded.

SoumyaRaikwar commented:
@yurishkuro Thanks for the clear direction! Let me make sure I understand the scope:

**What I need to do:**
1. Update `CompareTraces` to work with OTLP traces instead of Jaeger format
2. Convert existing test fixtures to OTLP
3. Add new fixtures for OTLP-specific features (scope attributes & span links)
4. Refactor `testGetTrace` to run fixture-based nested tests
5. Add exclusion logic for v1 backends (similar to how query tests work)

**A few questions before I dive in:**
1. Is there existing OTLP comparison code I should reference, or should I build this from scratch?
2. For fixtures - should they be OTLP JSON format (what the collector receives)?
3. You mentioned being skeptical about scope/link attributes getting lost - should I investigate the OTLP→Jaeger translator first to see what's actually preserved?

I'll start by looking at the translator and updating `CompareTraces` since that's the foundation for everything else.
